File: algorithms/0016/main.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number}
 */
var threeSumClosest = function(nums, target) {
  if (nums.length <= 3) {
    return nums.reduce((acc, cur) => {
      acc += cur
      return acc
    }, 0)
  }
  nums = nums.sort((a, b) => a - b)
  let result = Number.POSITIVE_INFINITY

  for (let i = 0; i < nums.length; i++) {
    // 减少不必要的循环
    if (i > 0 && nums[i] === nums[i - 1]) {
      continue
    }
    let l = i + 1
    let r = nums.length - 1
    while (l < r) {
      const sum = nums[i] + nums[l] + nums[r]
      if (sum === target) {
        return sum
      }
      const tmp = Math.abs(result - target)
      if (sum < target) {
        if (target - sum < tmp) {
          result = sum
        }
        l++
      } else if (sum > target) {
        if (sum - target < tmp) {
          result = sum
        }
        r--
      }
    }
  }
  return result
}

module.exports = threeSumClosest

File: algorithms/0016/main.go
package p0016

import (
	"math"
	"sort"
)

func threeSumClosest(nums []int, target int) int {
	if len(nums) <= 3 {
		sum := 0
		for _, num := range nums {
			sum += num
		}
		return sum
	}
	sort.Ints(nums)
	result := math.MaxInt64
	for i := 0; i < len(nums); i++ {
		if i > 0 && nums[i] == nums[i-1] {
			continue
		}
		l, r := i+1, len(nums)-1
		for l < r {
			sum := nums[i] + nums[l] + nums[r]
			abs := abs(result - target)
			switch {
			case sum < target:
				if target-sum < abs {
					result = sum
				}
				l++
			case sum > target:
				if sum-target < abs {
					result = sum
				}
				r--
			}
		}
	}
	return result
}

func abs(n int) int {
	if n < 0 {
		return -n
	}
	return n
}
